src/Components/Analyzers/src/InternalUsageAnalyzer.cs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.Extensions.Internal
+{
+    internal class InternalUsageAnalyzer
+    {
+        private readonly Func<ISymbol, bool> _isInternalNamespace;
+        private readonly Func<ISymbol, bool> _hasInternalAttribute;
+        private readonly DiagnosticDescriptor _descriptor;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="InternalUsageAnalyzer" />. The creator should provide delegates to help determine whether
+        /// a given symbol is internal or not, and a <see cref="DiagnosticDescriptor" /> to create errors.
+        /// </summary>
+        /// <param name="isInInternalNamespace">The delegate used to check if a symbol belongs to an internal namespace.</param>
+        /// <param name="hasInternalAttribute">The delegate used to check if a symbol has an internal attribute.</param>
+        /// <param name="descriptor">
+        /// The <see cref="DiagnosticDescriptor" /> used to create errors. The error message should expect a single parameter
+        /// used for the display name of the member.
+        /// </param>
+        public InternalUsageAnalyzer(Func<ISymbol, bool> isInInternalNamespace, Func<ISymbol, bool> hasInternalAttribute, DiagnosticDescriptor descriptor)
+        {
+            _isInternalNamespace = isInInternalNamespace ?? new Func<ISymbol, bool>((_) => false);
+            _hasInternalAttribute = hasInternalAttribute ?? new Func<ISymbol, bool>((_) => false);
+            _descriptor = descriptor ?? throw new ArgumentNullException(nameof(descriptor));
+        }
+
+        public void Register(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.RegisterSyntaxNodeAction(AnalyzeNode,
+                SyntaxKind.SimpleMemberAccessExpression,
+                SyntaxKind.ObjectCreationExpression,
+                SyntaxKind.ClassDeclaration,
+                SyntaxKind.Parameter);
+        }
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            switch (context.Node)
```

Comment:
I find it pretty amazing that (or *if*) this sort of thing doesn't impact compile times noticeably. Is it basically because the normal C# compilation process already looks like this internally but is 1000x bigger?

Reply:
The compiler is already pretty smart, and all of the data we're accessing is cached and used already. Basically we register callbacks and then operate on the nodes the compiler gives us. We don't really get to decide what nodes the compiler produces which is why it's efficient.

Here's some data from the MVC template that has more analyzers: `dotnet msbuild /t:rebuild /p:ReportAnalyzer=true /bl`

```txt
Using shared compilation with compiler from directory: C:\Program Files\dotnet\sdk\3.0.100-preview8-013528\Roslyn\bincore
Microsoft (R) Visual C# Compiler version 3.3.0-beta2-19376-02 (fc001df5)
Copyright (C) Microsoft Corporation. All rights reserved.
Total analyzer execution time: 0.088 seconds.
NOTE: Elapsed time may be less than analyzer execution time because analyzers can run concurrently.
Time (s)    %   Analyzer
0.080   91   Microsoft.AspNetCore.Analyzers, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60
0.080   91      Microsoft.AspNetCore.Analyzers.StartupAnalzyer
0.005    6   Microsoft.AspNetCore.Mvc.Analyzers, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60
0.002    1      Microsoft.AspNetCore.Mvc.Analyzers.TagHelpersInCodeBlocksAnalyzer
0.001    1      Microsoft.AspNetCore.Mvc.Analyzers.TopLevelParameterNameAnalyzer
0.001    1      Microsoft.AspNetCore.Mvc.Analyzers.AvoidHtmlPartialAnalyzer
0.001    1      Microsoft.AspNetCore.Mvc.Analyzers.AttributesShouldNotBeAppliedToPageModelAnalyzer
0.002    2   Microsoft.AspNetCore.Components.Analyzers, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60
0.002    1      Microsoft.AspNetCore.Components.Analyzers.ComponentParameterAnalyzer
<0.001   <1      Microsoft.AspNetCore.Components.Analyzers.ComponentParameterUsageAnalyzer
```

The analyzer runtime was less than 9ms. 

Obviously this doesn't include the analyzer getting worked on here, but these analyzers do similar things.

--- 

Ultimately I'd want to make some improvements to this code overall. The approach us using Syntax + Semantic model is the *old bad way* of doing things, and they now have something purpose built for what we need.